Add note date range logic

louwie17 opened the pull request:
Partially addresses #4789 

Defines a set of ranges for the store age the can be used for displaying Notes, this includes an upper range, which should prevent newer notes from showing up in older stores. 
There will be follow up PRs to update the rest of the notes, as part of this PR I updated a couple different ones that directly overlapped with a defined range.

### Detailed test instructions:

- On a new store with the onboarding flow finished and having the WC Admin test helper plugin installed ([latest release here](https://github.com/woocommerce/woocommerce-admin-test-helper/releases/tag/v0.2.0))
- Make sure the `You're invited to share your experience` note is not displayed (givingFeedback)
- Update the `woocommerce_admin_install_timestamp` option with a timestamp older then 4 weeks - (Ex: `1617888471` (April 8th)) through the test helper **Tools > WCA Test Helper > Options**
- Go to **Tools > WCA Test Helper > Tools** and run the wc_admin_daily job
- Go to **WooCommerce > Home** and notice that the `You're invited to share your experience` note is not displayed
- Update the `woocommerce_admin_install_timestamp` option again with a timestamp between 7 days and 4 weeks (28 days).
- Trigger the `wc_admin_daily` job again
- Go to **WooCommerce > Home** and notice that the `You're invited to share your experience` note is now displayed
- When updating the store to older to 4 weeks again should keep the note displayed (until dismissed).

<!-- 
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance: to readme.txt under the "unreleased" list at the top of the changelog. If none exists, please add it. Also include PR number.

If changes pertain to a package, also update CHANGELOG.md in the package's folder in a similar manner.--->

louwie17 committed:
Add notes helper method for date ranges

louwie17 committed:
Update two of the notes with date ranges

louwie17 committed:
Add tests and fix range logic

louwie17 committed:
Add changelog

louwie17 commented on src/WCAdminHelper.php:
```diff
@@ -46,4 +68,20 @@ public static function is_wc_admin_active_for( $seconds ) {
+	 * @param string $range range available in WC_ADMIN_STORE_AGE_RANGES.
+	 * @return bool Whether or not WooCommerce admin has been active within the range.
+	 */
+	public static function is_wc_admin_active_in_date_range( $range ) {
```
Curious if there is a way to add the accepted string types to the `$range` method here, instead of any string?

adrianduffell replied:
Not yet sadly, enums don't come until PHP 8.1 https://stitcher.io/blog/php-enums

In the meantime you could do a check like `array_key_exists( $range, self::WC_ADMIN_STORE_AGE_RANGES )` and throw an `\InvalidArgumentException` if it fails.

louwie17 replied:
Thanks that's a nice way as well (I look forward to enums :) ), fixed as part of this commit: https://github.com/woocommerce/woocommerce-admin/pull/6969/commits/aa2cb35f30506114f31f3adafa955b9e3885bea2

louwie17 committed:
Update comments

louwie17 commented on src/WCAdminHelper.php:
```diff
@@ -18,6 +18,28 @@ class WCAdminHelper {
 	 */
 	const WC_ADMIN_TIMESTAMP_OPTION = 'woocommerce_admin_install_timestamp';
 
```
Ranges as defined by Pedro here: https://github.com/woocommerce/woocommerce-admin/issues/4789#issuecomment-762752544

becdetat commented:
> Update the woocommerce_admin_install_timestamp option 

The latest version of the [test helper](https://github.com/woocommerce/woocommerce-admin-test-helper/releases) includes a tool to set the install timestamp by entering the install date, could save some time.

adrianduffell commented on tests/wc-admin-helper.php:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	/**
+	 * @dataProvider range_provider
```
🎉 This data provider is nice!

becdetat replied:
Agreed, this is the first time I've seen it :-)

louwie17 replied:
Yeah it took me a little while to find an example, as it is not used often, but nice indeed :)

adrianduffell commented on src/Notes/NoteTraits.php:
```diff
@@ -25,6 +25,16 @@ public static function wc_admin_active_for( $seconds ) {
+	 * @param string $range range available in WC_ADMIN_STORE_AGE_RANGES.
+	 * @return bool Whether or not WooCommerce admin has been active within the range.
+	 */
+	public static function is_wc_admin_active_in_date_range( $range ) {
```
This could be `private` since it is only ever called internally.

adrianduffell reviewed (changes requested):
@louwie17 Nice work on this and thanks for all the tests. I just wasn't sure about the `week-2-4` ranges and made some suggestions.

becdetat commented:
In the WC Admin Test Helper plugin there's also a tool I just discovered to run the `wc_admin_daily` task, without needing to install crontrol. Kudos to @moon0326 👍

louwie17 commented on src/WCAdminHelper.php:
```diff
@@ -18,6 +18,28 @@ class WCAdminHelper {
+			'end'   => WEEK_IN_SECONDS,
+		),
+		'week-2-4'  => array(
+			'start' => WEEK_IN_SECONDS,
```
Yeah the naming might be confusing week, I think I will update it to `week-1-4` as it is suppose to be after the first week till the end of the 4th week.

louwie17 replied:
I updated the name instead to `week-1-4`.

louwie17 commented on src/WCAdminHelper.php:
```diff
@@ -18,6 +18,28 @@ class WCAdminHelper {
+		),
+		'week-2-4'  => array(
+			'start' => WEEK_IN_SECONDS,
+			'end'   => MONTH_IN_SECONDS,
```
Yeah I originally had it as this, then changed it to `MONTH_IN_SECONDS`, but might change it back to what you suggested.
I did this to align it perfectly with the next range, and wasn't sure if the 2 day gap in-between could cause problems (it shouldn't, if we depend only on a single range).

louwie17 replied:
Updated in: https://github.com/woocommerce/woocommerce-admin/pull/6969/commits/aa2cb35f30506114f31f3adafa955b9e3885bea2

louwie17 committed:
Update range name and added exception if invalid range passed in.

louwie17 committed:
Fix php unit tests

louwie17 committed:
Fix another test

louwie17 commented:
@adrianduffell and @becdetat thanks for the review and suggestions, I updated the PR and updated the description to make use of the test helper plugin :) It should be good for a re-review.

adrianduffell reviewed (approved):
This looks good and tested well 🚀  

Just noting, I think the `GivingFeedback` note is currently disabled from PR https://github.com/woocommerce/woocommerce-admin/pull/6802? I needed to re-enable it to follow the test instructions.

louwie17 commented:
> Just noting, I think the GivingFeedback note is currently disabled from PR #6802? I needed to re-enable it to follow the test instructions.

Thanks, and my bad, not sure how I missed that, I will have to double check that on my end (just in case).
